### PR TITLE
Add per-file-ignores

### PR DESCRIPTION
**Pull request by LiamPattinson**
Fixes https://github.com/PlasmaFAIR/fortitude/issues/6

Allows the user to specify `--per-file-ignores` and `--extend-per-file-ignores` to switch off rules/categories for specific files:

```bash
fortitude check --per-file-ignores=myfile.f90:typing,myotherfile.f90:S001
```

Using `--extend-per-file-ignores` on the command line adds to the list in the config file, while `--per-file-ignores` overwrites it.

As we're just using `serde` to handle the config file vs CLI problem, the `.toml` interface is a bit clunky:

```toml
[check]
per-file-ignores = [
  "myfile.f90:typing",
  "myotherfile.f90:S001",
]
```

Ideally we would have a separate interface like:

```toml
[check.per-file-ignores]
"myfile.f90" = ["typing"]
"myotherfile.f90" = ["S001"] 
```

I'd suggest raising a new issue for this if this PR is merged.

---

I had to copy quite a lot of new Ruff machinery over to get this working, and while I've tried it out on some local files, I haven't written any tests yet. I'll leave this as a draft until those are done.

### CODE REVIEW

File: fortitude/src/check.rs [hunk 1]
```diff
@@ -640,6 +660,19 @@ pub fn check(args: CheckArgs, global_options: &GlobalConfigArgs) -> Result<ExitC
         extend_select: args.extend_select.unwrap_or(file_settings.extend_select),
     };
 
+    let mut per_file_ignores = args
+        .per_file_ignores
+        .or(file_settings.per_file_ignores)
+        .unwrap_or_default();
+    per_file_ignores.extend(
```
Comment (ZedThree) [inline fortitude/src/check.rs:667]: Does this not work?

```suggestion
    let per_file_ignores = args
        .per_file_ignores
        .or(file_settings.per_file_ignores)
        .unwrap_or_default()
        .extend(
```

**Comment by ZedThree** [comment]
Sounds like we might need to have the split types for CLI and config file sooner rather than later.

My first thought on doing this was to filter the ruleset for each file beforehand (in the loop, but before we check), but this way looks pretty expedient and clean.

File: fortitude/src/check.rs (hunk 1 above)
Comment (LiamPattinson) [reply fortitude/src/check.rs]: I couldn't get that to work, but there was a similar iterator trick to avoid the mutable intermediate

**Comment by danielhollas** [comment]
Hi folks, 

this is a missing feature that's currently blocking us from using Fortitude in CI (ideally we'd like to just exclude all rules for certain files). Feel free to ping me once things start converging and I am happy to take this PR for a spin on our codebase.

Thanks for all the work!

**Comment by LiamPattinson** [comment]
@danielhollas I'm glad to hear that you're finding this useful, and I hope that by our next release we'll have enough configuration options for the most common use-cases. We're aiming to also add per-line rule ignores, the ability to exclude certain file patterns, and the ability to follow rules in the user's `.gitignore` files.

> ideally we'd like to just exclude all rules for certain files

I've just tested this, and the following seems to work:

`fortitude check --per-file-ignores=myfile.f90:ALL src/`

The file pattern can be finicky however. It will match all files with a certain file name (including with wildcards), but I can't get it to match files within a certain directory, which is something that you can do with `ruff`:

`fortitude check --per-file-ignores=./src/myfile.f90:ALL src/`  # Doesn't work for some reason

@ZedThree Would you be in favour of merging this now and raising a new issue for the directory-selection bug, or should I get that fixed first?

**Comment by ZedThree** [comment]
I think that needs to be fixed, `src/myfile.f90` should work, and `myfile.f90` should not match `src/myfile.f90` -- `**/myfile.f90` should match both `myfile.f90` and `src/myfile.f90`

**Comment by LiamPattinson** [comment]
I think I've gotten the directory selection to work, but it's strange in the tests -- it seems we need to specify `/tmp/.tmpDSFGJ725/` (or whatever) in the path to get `**/some_dir/*.f90` to work:

```
fortitude check --per-file-ignores=/tmp/.tmpSFGJ525/**/some_dir/*.f90 /tmp/.tmpSFGJ525
```

Outside of the tests however, the following seems to work just fine:

```
fortitude check --per-file-ignores=**/some_dir/*.f90 src
```

---

> I think that needs to be fixed, src/myfile.f90 should work, and myfile.f90 should not match src/myfile.f90 -- **/myfile.f90 should match both myfile.f90 and src/myfile.f90

Ruff's behaviour is that `myfile.f90` _should_ match `src/myfile.f90`. I could try to get this implemented, but it might take me a little while to work out exactly how the matching works and modify it accordingly.

**Comment by ZedThree** [comment]
> Ruff's behaviour is that `myfile.f90` _should_ match `src/myfile.f90`. I could try to get this implemented, but it might take me a little while to work out exactly how the matching works and modify it accordingly.

Ahh, this is [`.gitignore`-like](https://git-scm.com/docs/gitignore) behaviour:

> - If there is a separator at the beginning or middle (or both) of the pattern, then the pattern is relative to the directory level of the particular .gitignore file itself. Otherwise the pattern may also match at any level below the .gitignore level.
> 
> - If there is a separator at the end of the pattern then the pattern will only match directories, otherwise the pattern can match both files and directories.
> 
> - For example, a pattern doc/frotz/ matches doc/frotz directory, but not a/doc/frotz directory; however frotz/ matches frotz and a/frotz that is a directory (all paths are relative from the .gitignore file).

File: fortitude/tests/check.rs [hunk 2]
```diff
@@ -647,3 +647,216 @@ end program test
...
+    // Expect:
+    // - Don't overwrite config file
+    // - File types of foo and baz but no bar
+    // - No files with index 2
+    assert_cmd_snapshot!(Command::cargo_bin(BIN_NAME)?
+                         .args(["--config-file", config_file.as_os_str().to_string_lossy().as_ref()])
+                         .arg("check")
+                         .arg("--select=typing")
+                         .arg(ignores)
+                         .arg(path),
```
Comment (ZedThree) [inline fortitude/tests/check.rs:813]: Change the directory the command is run in:

```suggestion
    let ignores = format!(
        "--extend-per-file-ignores=**/double_nested/*.f90:implicit-typing",
    );

    apply_common_filters!();
    // Expect:
    // - Don't overwrite config file
    // - File types of foo and baz but no bar
    // - No files with index 2
    assert_cmd_snapshot!(Command::cargo_bin(BIN_NAME)?
                         .args(["--config-file", config_file.as_os_str().to_string_lossy().as_ref()])
                         .arg("check")
                         .arg("--select=typing")
                         .arg(ignores)
                         .current_dir(path),
```

Comment (LiamPattinson) [reply fortitude/tests/check.rs]: Thanks, that seems to have done it! I've done this and made some further simplifications in https://github.com/PlasmaFAIR/fortitude/pull/232/commits/960ea592c805799f1ddafaac01f9dab78ea5e8f2

**Pull request merged**